RFC-0030-Consolidate-Elastic-and-TorchX.md

```diff
@@ -0,0 +1,321 @@
+
+An interesting observation is that despite the amount of fragmentation in the tooling for PyTorch,
+when it comes to process launchers, most of these tools either directly invoke or are compatible with
+`torchrun`. This is the case with [Accelerate](https://huggingface.co/blog/pytorch-ddp-accelerate-transformers),
+[DeepSpeed](https://github.com/microsoft/DeepSpeed/blob/master/deepspeed/launcher/launch.py#L29), 
+[Megatron-LM](https://github.com/NVIDIA/Megatron-LM/blob/main/examples/run_text_generation_server_345M.sh#L17),
+and [SageMaker](https://sagemaker.readthedocs.io/en/stable/frameworks/pytorch/using_pytorch.html#launching-a-distributed-training-job).
+So there is something to be said about the acceptance and adoption of standards when it belongs in 
+the core pytorch library versus it being offered as a separate and optional install. 
+Having TorchX specs/APIs in core encourages the community to adopt and extend PyTorch-defined infrastructure
+standards. For the user, this means better less fragmentation, better portability and uniform UX. 
+For PyTorch, this means being able to leverage and dictate its runtime requirements and environment.
+ 
+
+## **Proposed Implementation**
+
+**TL;DR - Upstream TorchX (specs and sensible default impls only) to PyTorch under a new sub-module called `torch.x` and pull TorchElastic out of
+`torch.distributed` and merge it with `torch.x`**
+
+The diagram below depicts the proposed module move + upstream (optional merge of the `multiprocessing` module not shown).
+
+![`torch.x` Before and After](RFC-0030-assets/modules_before_after.png)
+
+**The following changes are proposed (refer to diagram above):**
+
+1. Move `torch.distribted.elastic.*` &rarr; `torch.x.elastic.*`
+2. Upstream `torchx.*` &rarr; `torch.x.*`. Here we have three options:
+     1. (Option 1) Upstream specs only:
+          1. `torchx.specs` (the interfaces and APIs)
+          2. Default implementations of said specs
+          3. Other implementations would remain in TorchX repo
+          4. PROS: Upstreams the minimal set of functionalities to have torch benefit from the topics mentioned in the [Motivation](#motivation) section 
+          5. CONS: 
+              1. Effectively splits the TorchX repo in two hence makes maintenance and CI/CD (testing) more complex
+              2. No built-in support for launching remote jobs
+     2. **[ PREFERRED ] (Option 2) Option 1 + Kubernetes and SLURM support**
+          1. PROS: Keeps the minimalistic approach of Option 1 while providing out-of-the-box for CSP agnostic remote scheduler support
+          2. CONS: TorchX still split into two repos
+         
+        > NOTE: The rest of the doc assumes Option 2 as this is the recommended option
+     3. (Option 3) Upstream everything in TorchX:
+          1. PROS: Makes maintenance and CI/CD simple since changes to the specs can be tested for downstream BC easily
+          2. CONS: bloat-torch
```

Comment:
Right so regardless of which option, the "extra-deps" of torchx would **never** be added as deps (even extra deps) of `torch`. **We won't mess with the `extras_require` of `torch`** (aka no `torch[batch]`) and instead would fail-fast with a warning prompting the user to install additional packages.